README.md
## FlScoreBar

A simple flutter score bar package that can be used in 2 states. `immutable` is the state that can be shown to the user.
And `mutable` is the state that can be changed by the user action. All widgets support Rtl direction.

## Usage

First of all, you must add the package to your `pubspec.yaml`:

```yaml
fl_score_bar: ^0.2.4
```

or enter this in your command line: 

```yaml
flutter pub add fl_score_bar
```


## Sample

![fl_score_bar](https://github.com/mobinsafaeian/FlScoreBar/blob/master/fl_score_bar.gif)

## FlScoreBar

FlScoreBar can be added to your widget tree like this:

```dart
  //immutable
  FlScoreBar( 
    title: 'score',
    maxScore: 5,
    score: 4.3,
    averageScoreColor: Colors.yellow,
    highScoreColor: Colors.green,
    lowScoreColor: Colors.red,
    textStyle: TextStyle(color: Colors.black),
  );
```

or 

```dart
  //mutable
  FlScoreBar.editable(
    title: 'score',
    maxScore: 5,
    score: 4.3,
    averageScoreColor: Colors.yellow,
    highScoreColor: Colors.green,
    lowScoreColor: Colors.red,
    textStyle: const TextStyle(color: Colors.black),
    onChanged: (value) {
      print('FlScoreBar updated value -> $value');
    },
   )
```

## IconScoreBar

IconScoreBar can be added to your widget tree like this:

```dart
  IconScoreBar(
    scoreIcon: Icons.star,
    iconColor: Colors.amber,
    score: 2.6,
    maxScore: 5,
    readOnly: false,
    onChanged: (value) {
      print('IconScoreBar updated value -> $value');
    },
   )
```



CHANGELOG.md
## 0.2.4 - [31/01/2022]

* Added Rtl support

## 0.2.3 - [29/01/2022]

* Complete documentations

## 0.2.2 - [29/01/2022]

* Added IconScoreBar that can used with custom icons.
* Added onChanged method to get notified when the score value is updated.

## 0.0.1 - [24/01/2022]

* Added ScoreBar that can be mutable/immutable.


lib/src/score_bar.dart
import 'package:fl_score_bar/src/colors.dart';
import 'package:fl_score_bar/src/custom_track_shape.dart';
import 'package:flutter/material.dart';

class FlScoreBar extends StatefulWidget {
  // title of the scoreBar (Optional)
  final String? title;
  // textStyle sets as a style of all Texts widget declared in this class (Optional)
  final TextStyle? textStyle;
  // current score (Required)
  final double score;
  // max score of the content (Optional - default value is 5)
  final int maxScore;
  // you can define custom colors to show better ui to user in different score values (Optional - default values sets in constructor)
  final Color highScoreColor;
  final Color lowScoreColor;
  final Color averageScoreColor;

  /// if true, user can edit the value
  /// else value is constant
  final bool _editable;

  /// onChange method to get the updated score value
  final void Function(double)? onChanged;

  /// This constructor sets _editable value to false and make slider immutable.
  /// [score] must be lower or equal than [maxScore] to make the widget work correctly.
  /// Also, [maxScore] must be between 2 and 8.
  const FlScoreBar({
    Key? key,
    this.title,
    required this.score,
    this.textStyle,
    this.highScoreColor = BaseColors.highScoreColor,
    this.lowScoreColor = BaseColors.lowScoreColor,
    this.averageScoreColor = BaseColors.averageScoreColor,
    this.maxScore = 5,
  })  : assert(score <= maxScore, 'score must be lower or equal than maxScore'),
        assert(maxScore < 8 && maxScore > 2,
            'maxScore must be lower than 8 and greater than 2'),
        _editable = false,
        onChanged = null,
        super(key: key);

  /// This named constructor sets _editable value to true and make slider visible to the user.
  /// [score] must be lower or equal than [maxScore] to make the widget work correctly.
  /// Also, [maxScore] must be between 2 and 8.
  const FlScoreBar.editable({
    Key? key,
    this.title,
    required this.score,
    this.textStyle,
    this.highScoreColor = BaseColors.highScoreColor,
    this.lowScoreColor = BaseColors.lowScoreColor,
    this.averageScoreColor = BaseColors.averageScoreColor,
    this.maxScore = 5,
    this.onChanged,
  })  : assert(score <= maxScore, 'score must be lower or equal than maxScore'),
        assert(maxScore < 8 && maxScore > 2,
            'maxScore must be lower than 8 and greater than 2'),
        _editable = true,
        super(key: key);

  @override
  _FlScoreBarState createState() => _FlScoreBarState();
}

class _FlScoreBarState extends State<FlScoreBar> {
  // Current score value - can be changed by slider
  late double _currentValue;

  @override
  void initState() {
    _currentValue = widget.score;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    // get bar color based on score value
    Color barColor = _getBarColor(context);

    return Container(
      padding: const EdgeInsets.fromLTRB(0, 23, 0, 0),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        mainAxisAlignment: MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          widget.title != null
              ? Text(
                  widget.title!,
                  style: widget.textStyle,
                )
              : Container(),
          const SizedBox(
            height: 4,
          ),
          // choose between immutable and mutable states
          widget._editable
              ? _editableScoreBarWidget(barColor) // mutable state
              : _scoreBarWidget(barColor) // immutable state
        ],
      ),
    );
  }

  /// Immutable ScoreBar widget
  Widget _scoreBarWidget(Color barColor) => Row(
        mainAxisSize: MainAxisSize.max,
        children: [
          for (int i = 0; i < widget.maxScore; i++)
            i != _currentValue.toInt()
                ? Flexible(
                    child: Container(
                      margin: const EdgeInsets.symmetric(horizontal: 1),
                      height: 6,
                      decoration: BoxDecoration(
                          color: i < _currentValue.toInt()
                              ? barColor
                              : Colors.transparent,
                          borderRadius: BorderRadius.circular(6)),
                    ),
                  )
                : Flexible(
                    child: Row(
                      mainAxisSize: MainAxisSize.max,
                      children: [
                        Expanded(
                          flex:
                              ((_currentValue - _currentValue.floorToDouble()) *
                                      10)
                                  .toInt(),
                          child: Container(
                            margin: const EdgeInsets.symmetric(horizontal: 1),
                            height: 6,
                            decoration: BoxDecoration(
                                color: barColor,
                                borderRadius: BorderRadius.circular(6)),
                          ),
                        ),
                        Expanded(
                          flex: 10 -
                              ((_currentValue - _currentValue.floorToDouble()) *
                                      10)
                                  .toInt(),
                          child: const SizedBox(
                            height: 6,
                          ),
                        )
                      ],
                    ),
                  ),
          const SizedBox(
            width: 8,
          ),
          Text(
            _currentValue.toStringAsFixed(2),
            style: widget.textStyle,
          )
        ],
      );

  /// Mutable ScoreBar Widget using Flutter Slider
  /// we create main score row widget based on [widget.score] value.
  /// Then, we had to put a [Slider] on the score widget to update the score
  /// value by user.
  Widget _editableScoreBarWidget(Color barColor) => Row(
        mainAxisSize: MainAxisSize.max,
        children: [
          Expanded(
            child: Stack(
              children: [
                Row(
                  mainAxisSize: MainAxisSize.max,
                  children: [
                    for (int i = 0; i < widget.maxScore; i++)
                      i != _currentValue.toInt()
                          ? Flexible(
                              child: Container(
                                margin:
                                    const EdgeInsets.symmetric(horizontal: 1),
                                height: 6,
                                decoration: BoxDecoration(
                                    color: i < _currentValue.toInt()
                                        ? barColor
                                        : Colors.transparent,
                                    borderRadius: BorderRadius.circular(6)),
                              ),
                            )
                          : Flexible(
                              child: Row(
                                mainAxisSize: MainAxisSize.max,
                                children: [
                                  Expanded(
                                    flex: ((_currentValue -
                                                _currentValue.floorToDouble()) *
                                            10)
                                        .toInt(),
                                    child: Container(
                                      margin: const EdgeInsets.symmetric(
                                          horizontal: 1),
                                      height: 6,
                                      decoration: BoxDecoration(
                                          color: barColor,
                                          borderRadius:
                                              BorderRadius.circular(6)),
                                    ),
                                  ),
                                  Expanded(
                                    flex: 10 -
                                        ((_currentValue -
                                                    _currentValue
                                                        .floorToDouble()) *
                                                10)
                                            .toInt(),
                                    child: const SizedBox(
                                      height: 6,
                                    ),
                                  )
                                ],
                              ),
                            ),
                  ],
                ),
                Positioned(
                  top: 0,
                  left: 0,
                  right: 0,
                  bottom: 0,
                  child: SliderTheme(
                    data: SliderThemeData(
                        trackShape:
                            CustomTrackShape()), // Defining custom TrackShape to remove the default padding.
                    child: Slider(
                      min: 0,
                      thumbColor: barColor,
                      activeColor: Colors
                          .transparent, // TrackShape colors must be transparent to make them invisible.
                      inactiveColor: Colors
                          .transparent, // TrackShape colors must be transparent to make them invisible.
                      max: widget.maxScore.toDouble(),
                      value: _currentValue,
                      onChanged: (value) {
                        /// Rebuild widget to change the current score value
                        /// and re-initialize the colors and score widget parts (only enabled in mutable state)
                        setState(() {
                          _currentValue = value;
                        });

                        /// call [widget.onChanged] method to notify that
                        /// the score value is updated
                        if (widget.onChanged != null) widget.onChanged!(value);
                      },
                    ),
                  ),
                )
              ],
            ),
          ),
          const SizedBox(
            width: 8,
          ),
          Text(
            _currentValue.toStringAsFixed(2),
            style: widget.textStyle,
          )
        ],
      );

  /// Return Color code based on score value
  Color _getBarColor(BuildContext context) {
    if (_currentValue / widget.maxScore >= 0.8) {
      return widget.highScoreColor;
    } else if (_currentValue / widget.maxScore >= 0.6) {
      return widget.averageScoreColor;
    } else {
      return widget.lowScoreColor;
    }
  }
}


lib/src/custom_clipper.dart
import 'package:flutter/material.dart';

/// IconClipper gets clip and direction values to clip the icon widget to make the half rating item.
class IconClipper extends CustomClipper<Rect> {
  /// This property defines the amount of clipping to clip the icon.
  final double clip;

  /// This property defines the way of clipping process based on direction(rtl/ltr).
  final bool isRtl;

  const IconClipper({
    required this.clip,
    required this.isRtl,
  });

  /// Create a [Rect] on widget ViewPort to clip the widget
  /// based on [clip] value
  @override
  Rect getClip(Size size) => isRtl
      ? Rect.fromLTRB(
          size.width,
          0,
          size.width - (size.width * clip),
          size.height,
        )
      : Rect.fromLTRB(
          0,
          0,
          size.width * clip,
          size.height,
        );

  @override
  bool shouldReclip(covariant CustomClipper<Rect> oldClipper) => true;
}


lib/src/icon_score_bar.dart
import 'package:fl_score_bar/src/custom_clipper.dart';
import 'package:fl_score_bar/src/custom_track_shape.dart';
import 'package:flutter/material.dart';

class IconScoreBar extends StatefulWidget {
  /// IconData that used for RatingItem
  final IconData scoreIcon;

  /// RatingItem color
  final Color iconColor;

  /// Current score value
  /// initialize score value in [readOnly] mode
  final double score;

  /// Maximum score can't be greater than 8.
  /// and can't be lower than 2
  final int maxScore;

  /// If true, the user can't change the score value
  /// else, the user can define score value in [IconScoreBar]
  /// default value is false
  final bool readOnly;

  /// onChange method to get the updated score value
  final void Function(double)? onChanged;

  /// [score] must be lower or equal than [maxScore] to make the widget work correctly.
  /// Also, [maxScore] must be between 2 and 8.
  const IconScoreBar({
    Key? key,
    required this.scoreIcon,
    required this.iconColor,
    required this.score,
    required this.maxScore,
    this.readOnly = false,
    this.onChanged,
  })  : assert(score <= maxScore, 'score must be lower or equal than maxScore'),
        assert(maxScore < 8 && maxScore > 2,
            'maxScore must be lower than 8 and greater than 2'),
        super(key: key);

  @override
  _IconScoreBarState createState() => _IconScoreBarState();
}

class _IconScoreBarState extends State<IconScoreBar> {
  /// [currentValue] is the current score value
  /// It's value will change when the user changes the score value.
  late double currentValue;

  /// Current direction(rtl/ltr)
  /// We need this property to set it to the custom clipper for clipping score icon.
  late bool isRtl;

  @override
  void initState() {
    /// Initialize current value with [widget.score]
    currentValue = widget.score;
    super.initState();
  }

  /// First of all, we create [IconScoreBar] structure with 0 to [widget.maxScore]th RatingItem
  /// with 0.3 opacity.
  /// Then, we create main score row widget based on [widget.score] value.
  /// Then, we had to put an invisible [Slider] on the score widget to update
  /// score value by user.
  @override
  Widget build(BuildContext context) {
    /// Set current direction
    isRtl = Directionality.of(context) == TextDirection.rtl;

    return Stack(
      children: [
        Row(
          mainAxisSize: MainAxisSize.max,
          children: List.generate(
              widget.maxScore,
              (index) => Expanded(
                    child: Icon(
                      widget.scoreIcon,
                      color: widget.iconColor.withOpacity(0.3),
                      size: 48,
                    ),
                  )),
        ),
        Row(
          mainAxisSize: MainAxisSize.max,
          children: List.generate(
              widget.maxScore, (index) => _buildRatingItem(index)),
        ),
        Visibility(
          maintainSize: true,
          maintainInteractivity: true,
          maintainState: true,
          maintainAnimation: true,
          maintainSemantics: true,
          visible: false,
          child: SliderTheme(
            data: SliderThemeData(trackShape: CustomTrackShape()),
            child: Slider(
              min: 0,
              max: widget.maxScore.toDouble(),
              value: currentValue,
              inactiveColor: Colors.transparent,
              activeColor: Colors.transparent,
              onChanged: widget.readOnly
                  ? null
                  : (value) {
                      /// Rebuild widget to change the current score value
                      /// and re-initialize the [currentValue] and score widget parts
                      setState(() {
                        currentValue = value;
                      });

                      /// call [widget.onChanged] method to notify that
                      /// the score value is updated
                      if (widget.onChanged != null) widget.onChanged!(value);
                    },
            ),
          ),
        ),
      ],
    );
  }

  /// Returns [Icon] when [index] is lower than [currentValue].
  /// Returns [Icon] wrapped in [ClipRect] with [IconClipper] CustomClipper to
  /// clip the icon.
  /// Returns [Container] when the [index] is greater than [currentValue] to
  /// show nothing.
  /// All widgets wrapped in [Expanded] because of their [Row] parent. So they
  /// stick together with the same width size.
  Widget _buildRatingItem(int index) {
    if (index < currentValue.toInt()) {
      return Expanded(
        child: Icon(
          widget.scoreIcon,
          color: widget.iconColor,
          size: 48,
        ),
      );
    } else if (index == currentValue.toInt()) {
      return Expanded(
        child: ClipRect(
            clipper: IconClipper(
              isRtl: isRtl,
              clip: currentValue - currentValue.toInt(),
            ),
            child: Icon(
              widget.scoreIcon,
              color: widget.iconColor,
              size: 48,
            )),
      );
    }
    return Expanded(child: Container());
  }
}


example/lib/main.dart
import 'package:fl_score_bar/fl_score_bar.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Score bar demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          // immutable
          const FlScoreBar(
            title: 'score',
            maxScore: 5,
            score: 4.3,
            averageScoreColor: Colors.yellow,
            highScoreColor: Colors.green,
            lowScoreColor: Colors.red,
            textStyle: TextStyle(color: Colors.black),
          ),
          // mutable
          FlScoreBar.editable(
            title: 'score',
            maxScore: 5,
            score: 4.3,
            averageScoreColor: Colors.yellow,
            highScoreColor: Colors.green,
            lowScoreColor: Colors.red,
            textStyle: const TextStyle(color: Colors.black),
            onChanged: (value) {
              print('FlScoreBar updated value -> $value');
            },
          ),
          // Icon Score Bar
          IconScoreBar(
            scoreIcon: Icons.star,
            iconColor: Colors.amber,
            score: 2.6,
            maxScore: 5,
            readOnly: false,
            onChanged: (value) {
              print('IconScoreBar updated value -> $value');
            },
          )
        ],
      ),
    );
  }
}
